### PR TITLE
Dev mode - tenant endpoints for testing against localhost

### PR DESCRIPTION
**Pull request by ClaudeROSSI**
I could declare tenant endpoints allowing to test the mobile app against a server running locally:
Strange enough but the solution is different depending on the target platform.

- With IOS, the connection to http://127.0.01 works properly
- With ANDROID, I had to add some configuration files and target http://10.0.2.2 instead

In both cases the ev-server port is expected to be 8020

I also noticed some instability on ANDROID
- The "**Picker**" control for selecting the tenant endpoints was crashing the app. I did some changes and could get rid of some lint warnings related to the current picker selection
- The **getTenantLogo(**) was throwing an exception which was not handled property. This has been changed as well

### CODE REVIEW

File: src/screens/tenants/CreateTenantDialog.tsx [hunk 1]
```diff
@@ -46,24 +47,27 @@ export default class CreateTenantDialog extends React.Component<Props, State> {
...
   public constructor(props: Props) {
     super(props);
-    this.tenantEndpointClouds = Configuration.ENDPOINT_CLOUDS;
+    let tenantEndpointClouds: EndpointCloud[];
+    if (__DEV__) {
+      tenantEndpointClouds = Configuration.ENDPOINT_CLOUDS_DEV;
+    } else {
+      tenantEndpointClouds = Configuration.ENDPOINT_CLOUDS;
+    }
+    const newTenantEndpointCloudId = tenantEndpointClouds.find((tenantEndpointCloud) => tenantEndpointCloud.id === 'scp').id;
```
Comment (AlixH) [inline src/screens/tenants/CreateTenantDialog.tsx:56]: Maybe add a chaining operator before the .id ? I think find returns undefined if nothing is found

Comment (ClaudeROSSI) [reply src/screens/tenants/CreateTenantDialog.tsx]: thanks. Done!

File: src/config/Configuration.tsx [hunk 2]
```diff
@@ -15,6 +15,12 @@ export default class Configuration {
     { id: 'scpqa', name: 'SAP Cloud Platform QA', endpoint: Configuration.SCP_REST_ENDPOINT_QA }
   ];
 
+  public static readonly ENDPOINT_CLOUDS_DEV: EndpointCloud[] = [
+    ...Configuration.ENDPOINT_CLOUDS,
+    { id: 'localhost:8020', name: '127.0.0.1:8020', endpoint: 'http://127.0.0.1:8020' },
+    { id: 'android:localhost:8020', name: '10.0.2.2:8020', endpoint: 'http://10.0.2.2:8020' }
+  ];
```
Comment (jerome-benoit) [inline src/config/Configuration.tsx:22]: The setup you put in there is specific to your local development environment and can't be hardcoded. You need to introduce a configuration file that permit to define custom endpoints for development purpose.

File: android/app/src/main/res/xml/network_security_config.xml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<network-security-config>
+    <domain-config cleartextTrafficPermitted="true">
+        <domain includeSubdomains="true">10.0.2.2</domain>
```
Comment (jerome-benoit) [inline android/app/src/main/res/xml/network_security_config.xml:4]: We can't ship that file in the production build with a hardcoded IP address ;) 
It must be: 
+ per dev env (template)
+ tied to the development build only.

File: src/screens/tenants/CreateTenantDialog.tsx [hunk 4]
```diff
@@ -46,24 +47,27 @@ export default class CreateTenantDialog extends React.Component<Props, State> {
...
   public constructor(props: Props) {
     super(props);
-    this.tenantEndpointClouds = Configuration.ENDPOINT_CLOUDS;
+    let tenantEndpointClouds: EndpointCloud[];
+    if (__DEV__) {
+      tenantEndpointClouds = Configuration.ENDPOINT_CLOUDS_DEV;
+    } else {
+      tenantEndpointClouds = Configuration.ENDPOINT_CLOUDS;
+    }
+    const newTenantEndpointCloudId = tenantEndpointClouds.find((tenantEndpointCloud) => tenantEndpointCloud.id === 'scp')?.id;
```
Comment (jerome-benoit) [inline src/screens/tenants/CreateTenantDialog.tsx:56]: I do not think the filtering on 'scp' is necessary. And you like to put the result of helpers in intermediate variables ;) I'm not very fond of it, I do not think that makes the code easier to read, especially with JS or TS that permits to use arrow function and the like to directly use treatment result. That's just nitpicking ;)

**Review by jerome-benoit — CHANGES_REQUESTED** [review]

File: src/screens/tenants/CreateTenantDialog.tsx (hunk 4 above)
Comment (ClaudeROSSI) [reply src/screens/tenants/CreateTenantDialog.tsx:56]: you're right ... 'scp' is just the preferred endpoint ... so initializing the initial state using a constant is sufficient and straightforward.
I'll change it

File: android/app/src/main/res/xml/network_security_config.xml (hunk 3 above)
Comment (ClaudeROSSI) [reply android/app/src/main/res/xml/network_security_config.xml:4]: Right. This is the reason why this is a  DRAFT PR. 
I would need your advise on how to proceed here.

File: src/screens/tenants/CreateTenantDialog.tsx (hunk 4 above)
Comment (ClaudeROSSI) [reply src/screens/tenants/CreateTenantDialog.tsx]: done

File: android/app/src/main/res/xml/network_security_config.xml (hunk 3 above)
Comment (ClaudeROSSI) [reply android/app/src/main/res/xml/network_security_config.xml]: Configuration has been moved to:
.../android/app/src/debug/res/xml/network_security_config.xml

I'm not sure this is sufficient.

File: assets/config-template.json [hunk 5]
```diff
@@ -6,9 +6,9 @@
       "endpoint": "http://127.0.0.1:8020"
     },
     {
-      "id": "localhost:8020",
-      "name": "localhost:8020",
-      "endpoint": "http://localhost:8020"
+      "id": "10.0.2.2::8020",
+      "name": "android-local:8020",
+      "endpoint": "http://10.0.2.2:8020"
```
Comment (ClaudeROSSI) [inline assets/config-template.json:11]: @jerome-benoit : using localhost does not work when testing with the Android emulator.
According to the documentation we have to use:
> 10.0.2.2 | Special alias to your host loopback interface

**Comment by ClaudeROSSI** [comment]
One remark here for testing the mobile app against localhost.

The config file gets load thanks to:
> import ConfigurationData from '../../assets/config.json';

This file only exists when created manually (based on the provided template) and declares additional CLOUD ENDPOINTS.
I'm not sure loading a configuration file using the "import" statement is something recommended. do you?

Thanks,
Claude

**Comment by ClaudeROSSI** [comment]
hi,

> npm run pod:install

does not run on my side anymore! (even with master-qa).

I'm getting:

> [!] CocoaPods could not find compatible versions for pod "Stripe":
>  In snapshot (Podfile.lock):
    Stripe (= 21.6.0, ~> 21.6.0)

>  In Podfile:
    stripe-react-native (from `../node_modules/@stripe/stripe-react-native`) was resolved to 0.1.4, which depends on
      Stripe (~> 21.6.0)

> None of your spec sources contain a spec satisfying the dependencies: `Stripe (= 21.6.0, ~> 21.6.0), Stripe (~> 21.6.0)`.

Any idea?
Thanks

**Comment by Ninjeneer** [comment]
> hi,
> 
> > npm run pod:install
> 
> does not run on my side anymore! (even with master-qa).
> 
> I'm getting:
> 
> > [!] CocoaPods could not find compatible versions for pod "Stripe":
> > In snapshot (Podfile.lock):
> > Stripe (= 21.6.0, ~> 21.6.0)
> 
> > In Podfile:
> > stripe-react-native (from `../node_modules/@stripe/stripe-react-native`) was resolved to 0.1.4, which depends on
> > Stripe (~> 21.6.0)
> 
> > None of your spec sources contain a spec satisfying the dependencies: `Stripe (= 21.6.0, ~> 21.6.0), Stripe (~> 21.6.0)`.
> 
> Any idea?
> Thanks

Running npm install another time fixed it for me

**Comment by AlixH** [comment]
Please do
npm clean-install
pod install in ios/ folder

From: Loan Alouache ***@***.***>
Date: Wednesday, 7 July 2021 at 12:47
To: sap-labs-france/ev-mobile ***@***.***>
Cc: HUMBERT, Alix ***@***.***>, Review requested ***@***.***>
Subject: Re: [sap-labs-france/ev-mobile] Dev mode - tenant endpoints for testing against localhost (#348)

hi,

npm run pod:install

does not run on my side anymore! (even with master-qa).

I'm getting:

[!] CocoaPods could not find compatible versions for pod "Stripe":
In snapshot (Podfile.lock):
Stripe (= 21.6.0, ~> 21.6.0)

In Podfile:
stripe-react-native (from ..***@***.***/stripe-react-native) was resolved to 0.1.4, which depends on
Stripe (~> 21.6.0)

None of your spec sources contain a spec satisfying the dependencies: Stripe (= 21.6.0, ~> 21.6.0), Stripe (~> 21.6.0).

Any idea?
Thanks

Running npm install another time fixed it for me

File: .vscode/settings.json [hunk 6]
```diff
@@ -1,3 +1,2 @@
 {
-  "workbench.colorTheme": "Solarized Dark"
```
Comment (ClaudeROSSI) [inline .vscode/settings.json]: I would like to remove and gitignore that file to make sure I can have a different colorTheme on my side while reviewing.
@LucasBrazi06 , @AlixH , @Ninjeneer : Is it fine with you?

Comment (AlixH) [reply .vscode/settings.json]: I don't use VSCode ;)

Comment (LucasBrazi06) [reply .vscode/settings.json]: @ClaudeROSSI Okay
@AlixH Fired :-)

Comment (Ninjeneer) [reply .vscode/settings.json]: @ClaudeROSSI I don't know if ignoring the whole file is a good idea. We could use here some useful editor settings later.

However if you want to change your VSCode style, you can still install the Solarized Dark extension, and not adding it to the .vscode/settings.json

Comment (Ninjeneer) [reply .vscode/settings.json]: by the way, there was previously a useful setting before that fixes all eslint issues on save :
```
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  }
```

**Pull request merged**